autonat.go

```diff
@@ -50,62 +51,61 @@ type AmbientAutoNAT struct {
 	ctx  context.Context
 	host host.Host
 
-	getAddrs GetAddrs
-
-	mx     sync.Mutex
-	peers  map[peer.ID][]ma.Multiaddr
-	status NATStatus
-	addr   ma.Multiaddr
+	candidatePeers chan network.Conn
+	observations   chan autoNATResult
+	status         atomic.Value
 	// Reflects the confidence on of the NATStatus being private, as a single
 	// dialback may fail for reasons unrelated to NAT.
 	// If it is <3, then multiple autoNAT peers may be contacted for dialback
 	// If only a single autoNAT peer is known, then the confidence increases
 	// for each failure until it reaches 3.
-	confidence int
+	confidence  int
+	lastInbound time.Time
+	lastProbe   time.Time
 
 	emitUnknown event.Emitter
 	emitPublic  event.Emitter
 	emitPrivate event.Emitter
 }
 
-// NewAutoNAT creates a new ambient NAT autodiscovery instance attached to a host
-// If getAddrs is nil, h.Addrs will be used
-func NewAutoNAT(ctx context.Context, h host.Host, getAddrs GetAddrs) AutoNAT {
-	if getAddrs == nil {
-		getAddrs = h.Addrs
-	}
+type autoNATResult struct {
+	NATStatus
+	address ma.Multiaddr
+}
 
+// NewAutoNAT creates a new ambient NAT autodiscovery instance attached to a host
+func NewAutoNAT(ctx context.Context, h host.Host) AutoNAT {
```

Comment:
How are you getting the addresses to probe now? The `GetAddrs` component was necessary as autorelay may rewrite the addresses returned by the host based on autonat observations.